### PR TITLE
Prepare unix domain server

### PR DESCRIPTION
**Pull request by NekoCdr**
Prepare unix domain server without business logic. It response with requested query method (if QM is correct).

### CODE REVIEW

File: src/defines.h [hunk 1]
```diff
@@ -19,6 +19,21 @@ namespace tasknow {
 using D_size_t = int16_t;
 constexpr const std::ptrdiff_t BytesForSize{sizeof(D_size_t)};
 
+inline constexpr const int ErrorCode{-1};
+
+constexpr const char* DaemonSockPath{"/tmp/tasknowd.socket"};
```
Comment (Endilll) [inline src/defines.h]: Use `std::string_view`

File: src/serverd.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#include <sys/un.h>
+#include <unistd.h>
+
+namespace tasknow::serverd {
+
+auto init(
+    struct sockaddr_un* server_addr,
+    struct sockaddr_un* client_addr,
+    socklen_t addr_len,
+    const char* sock_path
```
Comment (Endilll) [inline src/serverd.cpp]: Use `std::string_view`

File: src/serverd.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    strcpy(static_cast<char*>(server_addr->sun_path), sock_path);
+
+    memset(client_addr, 0, addr_len);
+
+    if (std::remove(sock_path) == ErrorCode && errno != ENOENT) {
+        throw std::string_view{"Can't delete socket file"};
+    }
+}
+
+auto create_socket(int* server_sock) -> void
```
Comment (Endilll) [inline src/serverd.cpp]: Why in-out param? You can simply return the created socket

File: src/serverd.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        *client_sock,
+        &query_method,
+        BytesForSize,
+        0
+    ) == ErrorCode) {
+        throw std::string_view{"Can't send answer"};
+    }
+}
+
+auto serve(const char* sock_path, const int backlog_size) -> void
```
Comment (Endilll) [inline src/serverd.cpp]: Why not use `std::string_view` here as well?

Comment (Endilll) [inline src/serverd.cpp]: Accepting values as `const` is not wrong, but certainly unconventional. In a not so often case when you really try to change value of in-parameter without using it later, compiler should catch it, and issue a warning.

File: src/serverd.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    socklen_t* addr_len
+) -> void {
+    *client_sock = accept(
+        *server_sock,
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        reinterpret_cast<struct sockaddr*>(client_addr),
+        addr_len
+    );
+    if (*client_sock == ErrorCode) {
+        throw std::string_view{"Can't accept client"};
```
Comment (Endilll) [inline src/serverd.cpp]: If you add `using namespace std::literals;` to your headers, you'd be able to write this much shorter: `throw "Can't accept client"sv;`

File: src/serverd.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace tasknow::serverd {
+
+auto init(
+    struct sockaddr_un* server_addr,
+    struct sockaddr_un* client_addr,
+    socklen_t addr_len,
+    const char* sock_path
+) -> void;
+auto create_socket(int* server_sock) -> void;
```
Comment (Endilll) [inline src/serverd.h]: Can you sprinkle empty lines in-between functions to increase readability?

**Review by Endilll — CHANGES_REQUESTED** [review]

**Review by Endilll — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @NekoCdr)

___
*[`src/defines.h` line 25 at r2](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12#-NkvLSuC7F3mcm3G3zi6:-NkvLSuD-56JZ4rb6gJK:b-aoe53d) ([raw file](https://github.com/NekoCdr-pets/tasknow/blob/0316560f6f189a1e33603d1f8c82e0c784a6cc9f/src/defines.h#L25)):*
> ```objc
> inline constexpr const int ErrorCode{-1};
> 
> constexpr const std::string_view DaemonSockPath{"/tmp/tasknowd.socket"};
> ```

You can use `""sv` here, and use `auto` instead of `std::string_view` then.

**Review by Endilll — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @NekoCdr)

___
*[`src/main.cpp` line 22 at r2](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12#-NkvMaTv4wD8kYYKDuhM:-NkvMaTv4wD8kYYKDuhN:b-uzras6) ([raw file](https://github.com/NekoCdr-pets/tasknow/blob/0316560f6f189a1e33603d1f8c82e0c784a6cc9f/src/main.cpp#L22)):*
> ```cpp
>         tasknow::serverd::serve();
>     } catch (const std::string_view& error_msg) {
>         std::cout << error_msg << std::endl;
> ```

You can use `\n` instead of `std::endl`.

**Review by NekoCdr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @Endilll)

___
*[`src/main.cpp` line 22 at r2](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12#-NkvMaTv4wD8kYYKDuhM:-NkvQy4mBM-QMh-0kQaA:b-896fix) ([raw file](https://github.com/NekoCdr-pets/tasknow/blob/0316560f6f189a1e33603d1f8c82e0c784a6cc9f/src/main.cpp#L22)):*
<details><summary><i>Previously, Endilll (Vlad Serebrennikov) wrote…</i></summary><blockquote>

You can use `\n` instead of `std::endl`.
</blockquote></details>

Done.

**Review by NekoCdr — COMMENTED** [review]
> Dismissed @NekoCdr from a discussion.
> *[Reviewable](https://reviewable.io/reviews/NekoCdr-pets/tasknow/12)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @Endilll)

File: src/serverd.h [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+auto accept_client(
+    int* server_sock,
+    struct sockaddr_un* client_addr,
+    socklen_t* addr_len
+) -> int;
+
+auto get_peer_name(
+    int* client_sock,
+    struct sockaddr_un* client_addr,
```
Comment (Endilll) [inline src/serverd.h]: it's not a C code, so no need for elaborated type name (`struct`) here and everywhere

File: src/serverd.cpp [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+        };
+    }
+}
+
+auto accept_client(
+    int* server_sock,
+    struct sockaddr_un* client_addr,
+    socklen_t* addr_len
+) -> int {
+    int client_sock{};
```
Comment (Endilll) [inline src/serverd.cpp]: no need to declare this outside of the loop

File: src/serverd.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+auto create_socket() -> int;
+
+auto bind_socket(
+    int* server_sock,
+    struct sockaddr_un* server_addr,
+    socklen_t addr_len
+) -> void;
+
+auto listen_socket(int* server_sock, const int backlog_size) -> void;
```
Comment (Endilll) [inline src/serverd.h]: marking by-value parameters as `const` is not conventional

File: src/defines.h [hunk 10]
```diff
@@ -13,12 +13,28 @@
...
+#include <string_view>
 
 namespace tasknow {
 
 using D_size_t = int16_t;
 constexpr const std::ptrdiff_t BytesForSize{sizeof(D_size_t)};
 
+inline constexpr const int ErrorCode{-1};
+
+constexpr const std::string_view DaemonSockPath{"/tmp/tasknowd.socket"};
```
Comment (Endilll) [inline src/defines.h]: why is this not `inline`?

**Review by Endilll — CHANGES_REQUESTED** [review]

**Review by Endilll — APPROVED** [review]

**Pull request merged**